src/pydvl/value/semivalues.py

```diff
@@ -262,16 +272,30 @@ def compute_generic_semivalues(
 
             # Ensure that we always have n_submitted_jobs running
             try:
-                for _ in range(n_submitted_jobs - len(pending)):
-                    samples = tuple(islice(sampler_it, batch_size))
+                while len(pending) < n_submitted_jobs:
+                    samples = dict(islice(sampler_it, batch_size))
                     if len(samples) == 0:
                         raise StopIteration
 
-                    pending.add(
-                        executor.submit(
-                            _marginal, u=u, coefficient=correction, samples=samples
+                    # Filter out samples for indices that have already converged
+                    if skip_converged and len(done.converged) > 0:
```

Comment:
Can be rewritten in functional form 

```python
filtered_samples = samples.items() 
if skip_converged and len(done.converged) > 0:
  filtered_samples = filter(lambda t: done.converged[t[0]], filtered_samples)
filtered_samples = tuple(filtered_samples)
```

even one-liner possible (but difficult to read).